tests_e2e/tests/recover_network_interface/recover_network_interface.py

```diff
@@ -91,14 +91,46 @@ def run(self):
         # The script should bring the primary network interface down and use the agent to recover the interface. These
         # commands will bring the network down, so they should be executed on the machine using CSE instead of ssh.
         script = f"""
-        set -euxo pipefail
-        ifdown {ifname};
-        nic_state=$(nmcli -g general.state device show {ifname})
-        echo Primary network interface state before recovering: $nic_state
-        source /home/{self._context.username}/bin/set-agent-env;
-        pypy3 -c 'from azurelinuxagent.common.osutil.redhat import RedhatOSUtil; RedhatOSUtil().check_and_recover_nic_state({formatted_ifname})';
-        nic_state=$(nmcli -g general.state device show {ifname});
-        echo Primary network interface state after recovering: $nic_state
+        set -uxo pipefail
+        
+        # The 'ifdown' network script is used to bring the network interface down. For some distros, this script 
+        # executes nmcli commands which can timeout and return non-zero exit codes. Allow 3 retries in case 'ifdown' 
+        # returns non-zero exit code. This is the same number of retries the agent allows in DefaultOSUtil.restart_if
+        retries=3;
+        ifdown_success=false
+        while [ $retries -gt 0 ]
+        do
+            echo Attempting to bring network interface down with ifdown...
+            ifdown {ifname};
+            exit_code=$?
+            if [ $exit_code -eq 0 ]; then
+                echo ifdown succeeded
+                ifdown_success=true
+                break
+            fi
+            echo ifdown failed with exit code $exit_code, try again after 5 seconds...
+            sleep 5
+            ((retries=retries-1))
+        done
+        
+        # Verify the agent network interface recovery logic only if 'ifdown' succeeded
+        if ! $ifdown_success ; then 
+            # Fail the script if 'ifdown' command didn't succeed
```

Comment:
printing the err msg for the last retry would be useful for debug? Error codes may not give enough information

Reply:
Full error message will be in stderr since I'm using -x option. 

To check I ran the script locally, but replaced 'ifdown eth0' with a command that doesnt exist. This is the stderr:

./test.sh
++ retries=3
++ ifdown_success=false
++ '[' 3 -gt 0 ']'
++ echo Attempting to bring network interface down with ifdown...
Attempting to bring network interface down with ifdown...
**++ command-dne
./test.sh: line 11: command-dne: command not found
++ exit_code=127**
++ '[' 127 -eq 0 ']'
++ echo ifdown failed with exit code 127, try again after 5 seconds...
ifdown failed with exit code 127, try again after 5 seconds...
++ sleep 5
++ (( retries=retries-1 ))
++ '[' 2 -gt 0 ']'
++ echo Attempting to bring network interface down with ifdown...
Attempting to bring network interface down with ifdown...
++ command-dne
./test.sh: line 11: command-dne: command not found
++ exit_code=127
++ '[' 127 -eq 0 ']'
++ echo ifdown failed with exit code 127, try again after 5 seconds...
ifdown failed with exit code 127, try again after 5 seconds...
++ sleep 5
++ (( retries=retries-1 ))
++ '[' 1 -gt 0 ']'
++ echo Attempting to bring network interface down with ifdown...
Attempting to bring network interface down with ifdown...
++ command-dne
./test.sh: line 11: command-dne: command not found
++ exit_code=127
++ '[' 127 -eq 0 ']'
++ echo ifdown failed with exit code 127, try again after 5 seconds...
ifdown failed with exit code 127, try again after 5 seconds...
++ sleep 5
++ (( retries=retries-1 ))
++ '[' 0 -gt 0 ']'
++ false
++ exit 1